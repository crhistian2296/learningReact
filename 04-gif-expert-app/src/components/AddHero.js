import React, { useState } from "react";
import PropTypes from "prop-types";

export default function AddHero({ setHeroes }) {
    const [inputValue, setInputValue] = useState(""); //useState tiene valor undefined por defecto

    const handleOnChange = (e) => {
        setInputValue(e.target.value);
        // console.log(e.target.value);
    };

    const handleSubmit = (e) => {
        e.preventDefault();
        if (inputValue.trim().length > 0) {
            setHeroes((initialState) => [inputValue, ...initialState]);
            setInputValue("");
        } else console.error("input void, please enter a hero name before submit your request");

        // console.log("submit:", inputValue);
    };

    return (
        <form onSubmit={handleSubmit}>
            <input
                type="text"
                value={inputValue}
                // placeholder="Nuevo heroe"
                onChange={handleOnChange}
            />
        </form>
    );
}

//validacion de requrimiento de setHeroes como proptype
AddHero.propTypes = {
    setHeroes: PropTypes.func.isRequired,
};
