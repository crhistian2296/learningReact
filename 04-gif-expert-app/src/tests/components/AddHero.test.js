import { shallow } from "enzyme";
// import '@testing-library/jest-dom'
import React from "react";
import AddHero from "../../components/AddHero";

describe("Pruebas en <AddHero/>", () => {
    const setHeroes = jest.fn();
    //se define completamente el wrapper 2 veces para conservar la ayuda de autocompletado
    let wrapper = shallow(<AddHero setHeroes={setHeroes} />);
    beforeEach(() => {
        jest.clearAllMocks();
        wrapper = shallow(<AddHero setHeroes={setHeroes} />);
    });

    test("Debe mostrar el componente correctamente", () => {
        expect(wrapper).toMatchSnapshot();
    });

    test("Debe mostrar los cambios en el input", () => {
        const input = wrapper.find("input");
        const value = "Hola mundo";

        input.simulate("change", { target: { value } });
        //Se espera que el parrafo tenga el valor modificado del input
        expect(wrapper.find("p").text().trim()).toBe(value);
    });

    test("Debe evitar establecer una nueva categoria vacia", () => {
        wrapper.find("form").simulate("submit", { preventDefault() {} });
        // expect(setHeroes).not.toHaveBeenCalled();
        expect(setHeroes).not.toBeCalled();
    });

    //* Mock de la implementacion de un alert
    jest.spyOn(window, "alert").mockImplementation(() => {});

    test("Debe llamar un alert si se intenta un submit vacio", () => {
        wrapper.find("form").simulate("submit", { preventDefault() {} });
        // expect(setHeroes).not.toHaveBeenCalled();
        expect(window.alert).toBeCalled();
    });

    test("No Debe llamar un alert si se intenta un submit con contenido", () => {
        wrapper.find("input").simulate("change", { target: { value: "hola" } });
        wrapper.find("form").simulate("submit", { preventDefault() {} });
        // expect(setHeroes).not.toHaveBeenCalled();
        expect(window.alert).not.toHaveBeenCalled();
    });

    test("Debe llamara a setHeroes y limpiar la caja de texto", () => {
        wrapper.find("input").simulate("change", { target: { value: "hola" } });
        wrapper.find("form").simulate("submit", { preventDefault() {} });
        //fragmento que compureba si se llamo algun tipo de funcion
        expect(setHeroes).toHaveBeenCalledWith(expect.any(Function));
        expect(wrapper.find("input").prop("value").trim()).toBe("");
    });
});
