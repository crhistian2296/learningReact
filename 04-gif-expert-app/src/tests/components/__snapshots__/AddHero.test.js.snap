// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddHero/> Debe mostrar el componente correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  //parrafo para hacer testing
  <p>
     
     
  </p>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
